File: ge-develop/examples/es/hccl_tp/cpp/src/make_pfa_hcom_graph.cpp
/**
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */
#include "es_showcase.h"
#include "es_PromptFlashAttention.h"
#include "es_Reshape.h"
#include "es_BatchMatMul.h"
#include "es_HcomAllReduce.h"
#include "es_Cast.h"
#include "es_AddRmsNorm.h"
#include "utils.h"
#include <memory>
#include <vector>
#include <iostream>
#include "ge/ge_api.h"

using namespace ge;
using namespace ge::es;

namespace {
// Core graph building function
std::vector<es::EsTensorHolder> MakePfaHcomGraph(
    es::EsTensorHolder query, es::EsTensorHolder key, es::EsTensorHolder value,
    es::EsTensorHolder atten_mask, es::EsTensorHolder quant_scale2, es::EsTensorHolder quant_offset2,
    es::EsTensorHolder mm_x2, es::EsTensorHolder arn_x1, es::EsTensorHolder arn_gamma,
    EsGraphBuilder &graph_builder) {
  auto query_fp16 = Cast(query, DT_FLOAT16);
  auto key_fp16 = Cast(key, DT_FLOAT16);
  auto value_fp16 = Cast(value, DT_FLOAT16);
  auto atten_mask_fp16 = Cast(atten_mask, DT_FLOAT16);
  auto quant_scale2_fp16 = Cast(quant_scale2, DT_FLOAT16);
  auto quant_offset2_fp16 = Cast(quant_offset2, DT_FLOAT16);
  auto mm_x2_fp16 = Cast(mm_x2, DT_FLOAT16);
  auto arn_x1_fp16 = Cast(arn_x1, DT_FLOAT16);
  auto arn_gamma_fp16 = Cast(arn_gamma, DT_FLOAT16);
  auto pfa_output = PromptFlashAttention(
    query_fp16, key_fp16, value_fp16,
    nullptr, atten_mask_fp16, nullptr, nullptr,
    nullptr, nullptr, nullptr,
    quant_scale2_fp16, quant_offset2_fp16,
    8, 1.0f, 214748647, 0, "BSH", 8, 0, 1
  );

  auto reshape_output = Reshape(pfa_output, std::vector<int64_t>{2, 128, 512});
  auto mm_output = BatchMatMul(reshape_output, mm_x2_fp16);
  auto hcom_output = HcomAllReduce(mm_output, "sum", "hccl_world_group");
  auto cast_output = Cast(hcom_output, DT_FLOAT16);
  auto arn_output = AddRmsNorm(arn_x1_fp16, cast_output, arn_gamma_fp16);

  auto mm_output_fp32 = Cast(mm_output, DT_FLOAT);
  auto hcom_output_fp32 = Cast(hcom_output, DT_FLOAT);
  auto arn_output_fp32 = Cast(arn_output.y, DT_FLOAT);

  return {mm_output_fp32, hcom_output_fp32, arn_output_fp32};
}
}

namespace es_showcase {

int RunGraph(ge::Graph &graph, const std::vector<ge::Tensor> &inputs,
             const std::string &output_prefix) {
  ge::Utils::PrintTensorsToFile(inputs, "input");
  std::map<ge::AscendString, ge::AscendString> options;
  auto *s = new (std::nothrow) ge::Session(options);
  if (s == nullptr) {
    std::cout << "Global session not ready" << std::endl;
    return -1;
  }
  static uint32_t next = 0;
  const uint32_t graph_id = next++;
  auto ret = s->AddGraph(graph_id, graph);
  if (ret != ge::SUCCESS) {
    std::cout << "AddGraph failed" << std::endl;
    delete s;
    return -1;
  }
  std::vector<ge::Tensor> outputs;
  ret = s->RunGraph(graph_id, inputs, outputs);
  if (ret != ge::SUCCESS) {
    std::cout << "RunGraph failed" << std::endl;
    (void)s->RemoveGraph(graph_id);
    delete s;
    return -1;
  }
  (void)s->RemoveGraph(graph_id);
  ge::Utils::PrintTensorsToFile(outputs, output_prefix);
  delete s;
  return 0;
}

void MakePfaHcomGraphByEsAndDump() {
  std::unique_ptr<ge::Graph> graph = MakePfaHcomGraphByEs();
  graph->DumpToFile(ge::Graph::DumpFormat::kOnnx, ge::AscendString("make_pfa_hcom_graph"));
}

std::unique_ptr<ge::Graph> MakePfaHcomGraphByEs() {
  // 1. 创建图构建器
  auto graph_builder = std::make_unique<EsGraphBuilder>("MakePfaHcomGraph");

  // 2. 创建输入节点 - 使用FP32，在图内部转换为FP16
  auto query = graph_builder->CreateInput(0, "query", ge::DT_FLOAT, ge::FORMAT_ND, {2, 128, 512});
  auto key = graph_builder->CreateInput(1, "key", ge::DT_FLOAT, ge::FORMAT_ND, {2, 128, 512});
  auto value = graph_builder->CreateInput(2, "value", ge::DT_FLOAT, ge::FORMAT_ND, {2, 128, 512});
  auto atten_mask = graph_builder->CreateInput(3, "atten_mask", ge::DT_FLOAT, ge::FORMAT_ND, {2, 128, 128});
  auto quant_scale2 = graph_builder->CreateInput(4, "quant_scale2", ge::DT_FLOAT, ge::FORMAT_ND, {1});
  auto quant_offset2 = graph_builder->CreateInput(5, "quant_offset2", ge::DT_FLOAT, ge::FORMAT_ND, {1});
  auto mm_x2 = graph_builder->CreateInput(6, "mm_x2", ge::DT_FLOAT, ge::FORMAT_ND, {2, 512, 512});
  auto arn_x1 = graph_builder->CreateInput(7, "arn_x1", ge::DT_FLOAT, ge::FORMAT_ND, {2, 128, 512});
  auto arn_gamma = graph_builder->CreateInput(8, "arn_gamma", ge::DT_FLOAT, ge::FORMAT_ND, {512});

  // 3. 构建图
  auto outputs = MakePfaHcomGraph(query, key, value, atten_mask, quant_scale2, quant_offset2,
                                   mm_x2, arn_x1, arn_gamma, *graph_builder);

  // 4. 设置输出
  for (size_t i = 0; i < outputs.size(); ++i) {
    (void) graph_builder->SetOutput(outputs[i], i);
  }

  // 5. 构建图
  return graph_builder->BuildAndReset();
}

int MakePfaHcomGraphByEsAndRun() {
  std::unique_ptr<ge::Graph> graph = MakePfaHcomGraphByEs();
  std::vector<ge::Tensor> inputs;

  // 准备输入数据
  std::vector<float> query_data(2 * 128 * 512, 1.0f);
  std::vector<float> key_data(2 * 128 * 512, 1.0f);
  std::vector<float> value_data(2 * 128 * 512, 1.0f);
  std::vector<float> atten_mask_data(2 * 128 * 128, 0.0f);
  std::vector<float> quant_scale2_data(1, 1.0f);
  std::vector<float> quant_offset2_data(1, 0.0f);
  std::vector<float> mm_x2_data(2 * 512 * 512, 1.0f);
  std::vector<float> arn_x1_data(2 * 128 * 512, 1.0f);
  std::vector<float> arn_gamma_data(512, 1.0f);

  // 使用FP32类型创建输入（图内部会转换为FP16）
  inputs.push_back(*ge::Utils::StubTensor<float>(query_data, {2, 128, 512}));
  inputs.push_back(*ge::Utils::StubTensor<float>(key_data, {2, 128, 512}));
  inputs.push_back(*ge::Utils::StubTensor<float>(value_data, {2, 128, 512}));
  inputs.push_back(*ge::Utils::StubTensor<float>(atten_mask_data, {2, 128, 128}));
  inputs.push_back(*ge::Utils::StubTensor<float>(quant_scale2_data, {1}));
  inputs.push_back(*ge::Utils::StubTensor<float>(quant_offset2_data, {1}));
  inputs.push_back(*ge::Utils::StubTensor<float>(mm_x2_data, {2, 512, 512}));
  inputs.push_back(*ge::Utils::StubTensor<float>(arn_x1_data, {2, 128, 512}));
  inputs.push_back(*ge::Utils::StubTensor<float>(arn_gamma_data, {512}));

  return RunGraph(*graph, inputs, "PfaHcomTP");
}

}  // namespace es_showcase

File: ge-develop/examples/es/hccl_tp/python/src/make_pfa_hcom_graph.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
# ----------------------------------------------------------------------------
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ----------------------------------------------------------------------------

import os
import sys
import numpy as np

from ge.es import GraphBuilder
from ge.graph import Tensor, DumpFormat
from ge.graph.types import DataType, Format
from ge.ge_global import GeApi
from ge.session import Session
from ge.es.all import *

BATCH_SIZE = 2
SEQ_LEN = 128
HIDDEN_SIZE = 512


def build_pfa_hcom_graph():
    # 1、创建图构建器
    builder = GraphBuilder("MakePfaHcomGraph")

    # 2、定义输入配置 
    input_configs = [
        ("query", [BATCH_SIZE, SEQ_LEN, HIDDEN_SIZE]),
        ("key", [BATCH_SIZE, SEQ_LEN, HIDDEN_SIZE]),
        ("value", [BATCH_SIZE, SEQ_LEN, HIDDEN_SIZE]),
        ("atten_mask", [BATCH_SIZE, SEQ_LEN, SEQ_LEN]),
        ("quant_scale2", [1]),
        ("quant_offset2", [1]),
        ("mm_x2", [BATCH_SIZE, HIDDEN_SIZE, HIDDEN_SIZE]),
        ("arn_x1", [BATCH_SIZE, SEQ_LEN, HIDDEN_SIZE]),
        ("arn_gamma", [HIDDEN_SIZE])
    ]

    # 3、创建输入节点并转换为FP16
    inputs_fp32 = []
    inputs_fp16 = []
    for idx, (name, shape) in enumerate(input_configs):
        input_fp32 = builder.create_input(
            index=idx,
            name=name,
            data_type=DataType.DT_FLOAT,
            shape=shape
        )
        inputs_fp32.append(input_fp32)
        inputs_fp16.append(Cast(input_fp32, dst_type=DataType.DT_FLOAT16))

    query_fp16, key_fp16, value_fp16, atten_mask_fp16, quant_scale2_fp16, \
    quant_offset2_fp16, mm_x2_fp16, arn_x1_fp16, arn_gamma_fp16 = inputs_fp16

    # 4、构建计算图
    pfa_output = PromptFlashAttention(
        query_fp16,           # query
        key_fp16,             # key
        value_fp16,           # value
        None,                 # pse_shift（可选）
        atten_mask_fp16,      # atten_mask
        None,                 # actual_seq_lengths（可选）
        None,                 # actual_seq_lengths_kv（可选）
        None,                 # dequant_scale1（可选）
        None,                 # quant_scale1（可选）
        None,                 # dequant_scale2（可选）
        quant_scale2_fp16,    # quant_scale2
        quant_offset2_fp16,   # quant_offset2
        num_heads=8,          # 注意力头数
        scale_value=1.0,      # 缩放因子
        pre_tokens=214748647, # 预分配token数
        next_tokens=0,        # 下一轮token数
        input_layout="BSH",   # 输入布局：Batch, Sequence, Hidden
        num_key_value_heads=8,# KV头数
        sparse_mode=0,        # 稀疏模式
        inner_precise=1       # 内部精度
    )

    # Reshape变换
    reshape_shape = builder.create_vector_int64([2, 128, 512])
    reshape_output = Reshape(pfa_output, reshape_shape)

    # BatchMatMul矩阵乘法
    mm_output = BatchMatMul(reshape_output, mm_x2_fp16)

    # HcomAllReduce多卡通信
    hcom_output = HcomAllReduce(
        mm_output,
        reduction="sum",           
        group="hccl_world_group"   
    )

    # Cast为FP16
    cast_output = Cast(hcom_output, dst_type=DataType.DT_FLOAT16)

    # AddRmsNorm归一化
    arn_output = AddRmsNorm(arn_x1_fp16, cast_output, arn_gamma_fp16)

    # 5、将中间结果转换回FP32作为输出
    mm_output_fp32 = Cast(mm_output, dst_type=DataType.DT_FLOAT)
    hcom_output_fp32 = Cast(hcom_output, dst_type=DataType.DT_FLOAT)
    arn_output_fp32 = Cast(arn_output.y, dst_type=DataType.DT_FLOAT)

    # 6、设置图输出节点
    builder.set_graph_output(mm_output_fp32, 0)
    builder.set_graph_output(hcom_output_fp32, 1)
    builder.set_graph_output(arn_output_fp32, 2)

    # 7、构建图
    return builder.build_and_reset()


def dump_pfa_hcom_graph(graph):
    graph.dump_to_file(format=DumpFormat.kOnnx, suffix="make_pfa_hcom_graph")


def run_graph(graph, device_id="0") -> int:


    # 1. 初始化GE环境
    # 获取环境变量
    rank_id = os.environ.get("RANK_ID", None)
    rank_table_file = os.environ.get("RANK_TABLE_FILE", None)

    # 构建配置字典
    config = {
        "ge.exec.deviceId": str(device_id),
        "ge.graphRunMode": "0"  
    }

    # 添加HCCL相关配置
    if rank_id is not None and rank_table_file is not None:
        config["ge.exec.rankTableFile"] = rank_table_file
        config["ge.exec.rankId"] = rank_id
        print(f"[Info] 多卡模式 - RANK_ID: {rank_id}, RANK_TABLE_FILE: {rank_table_file}")
    else:
        print("[Info] 单卡模式 - 未检测到RANK_ID和RANK_TABLE_FILE环境变量")

    ge_api = GeApi()
    ret = ge_api.ge_initialize(config)
    if ret != 0:
        print(f"[Error] GE初始化失败，返回码: {ret}")
        return ret

    print(f"[Info] GE环境初始化成功 (Device ID: {device_id}, RANK_ID: {rank_id if rank_id else 'N/A'})")

    try:
        # 2. 创建Session
        session = Session()

        # 3. 添加图到Session
        graph_id = 1
        ret = session.add_graph(graph_id, graph)
        if ret != 0:
            print(f"[Error] 添加图失败，返回码: {ret}")
            return ret
        print(f"[Info] 图已添加到Session (Graph ID: {graph_id})")

        # 4. 准备输入数据
        input_data_configs = [
            ([BATCH_SIZE, SEQ_LEN, HIDDEN_SIZE], 1.0),       # query
            ([BATCH_SIZE, SEQ_LEN, HIDDEN_SIZE], 1.0),       # key
            ([BATCH_SIZE, SEQ_LEN, HIDDEN_SIZE], 1.0),       # value
            ([BATCH_SIZE, SEQ_LEN, SEQ_LEN], 0.0),           # atten_mask
            ([1], 1.0),                                      # quant_scale2
            ([1], 0.0),                                      # quant_offset2
            ([BATCH_SIZE, HIDDEN_SIZE, HIDDEN_SIZE], 1.0),   # mm_x2
            ([BATCH_SIZE, SEQ_LEN, HIDDEN_SIZE], 1.0),       # arn_x1
            ([HIDDEN_SIZE], 1.0)                             # arn_gamma
        ]

        # 创建Tensor对象
        inputs = []
        for shape, init_value in input_data_configs:
            data = np.full(shape, init_value, dtype=np.float32)
            tensor = Tensor(
                data.flatten().tolist(),
                None,
                DataType.DT_FLOAT,
                Format.FORMAT_ND,
                shape
            )
            inputs.append(tensor)

        print(f"[Info] 输入数据已准备，共{len(inputs)}个输入tensor")

        # 5. 运行图
        ret = session.run_graph(graph_id, inputs)
        print("[Info] 图运行成功！")
        for idx, tensor in enumerate(ret, start=1):
            print(f"Tensor{idx}详情：{tensor}")
        return 0

    except Exception as e:
        print(f"[Error] 执行过程中出错: {e}")
        import traceback
        traceback.print_exc()
        return -1

    finally:
        # 6. 清理GE环境
        print("[Info] 清理GE环境...")
        ge_api.ge_finalize()
        print("[Success] GE环境已清理")


if __name__ == "__main__":
    # 构建图
    graph = build_pfa_hcom_graph()

    # 先dump图（生成pbtxt文件用于可视化）
    dump_pfa_hcom_graph(graph)

    # 运行图（从命令行参数获取device_id，默认为"0"）
    device_id = sys.argv[1] if len(sys.argv) > 1 else "0"
    ret = run_graph(graph, device_id)
    sys.exit(ret)